File: ios/BareKitModule.m
#import <BareKit/BareKit.h>
#import <Foundation/Foundation.h>

#import "BareKitModule.h"

@interface BareKitModuleWorklet : NSObject

@end

@implementation BareKitModuleWorklet {
@public
  NSNumber *_id;
  BareKitModule *_module;
  BareWorklet *_worklet;
  BareIPC *_ipc;
}

- (_Nullable instancetype)initWithModule:(BareKitModule *)module
                                      id:(NSNumber *)id
                                filename:(NSString *)filename
                                  source:(NSString *)source
                               arguments:(NSArray<NSString *> *)arguments
                             memoryLimit:(nonnull NSNumber *)memoryLimit
                                  assets:(NSString *)assets {
  self = [super init];

  if (self) {
    _id = id;
    _module = module;

    BareWorkletConfiguration *options = [[BareWorkletConfiguration alloc] init];
    options.memoryLimit = memoryLimit.unsignedIntegerValue;
    options.assets = assets;

    _worklet = [[BareWorklet alloc] initWithConfiguration:options];

    NSData *decoded;

    if (source == nil) {
      decoded = nil;
    } else {
      decoded = [[NSData alloc] initWithBase64EncodedString:source options:0];
    }

    [_worklet start:filename source:decoded arguments:arguments];

    _ipc = [[BareIPC alloc] initWithWorklet:_worklet];
  }

  return self;
}

- (void)_suspend:(NSNumber *)linger {
  [_worklet suspendWithLinger:linger.intValue];
}

- (void)_resume {
  [_worklet resume];
}

- (void)_terminate {
  [_ipc close];
  [_worklet terminate];
}

@end

@implementation BareKitModule {
  int _id;
  NSMutableDictionary<NSNumber *, BareKitModuleWorklet *> *_worklets;
}

RCT_EXPORT_MODULE(BareKit)

- (_Nullable instancetype)init {
  self = [super init];

  if (self) {
    _id = 0;
    _worklets = [[NSMutableDictionary alloc] init];
  }

  return self;
}

- (void)invalidate {
  for (NSNumber *id in _worklets) {
    [_worklets[id] _terminate];
  }

  [_worklets removeAllObjects];
}

RCT_EXPORT_METHOD(start : (NSString *) filename
                  source : (NSString *) source
                  arguments : (NSArray *) arguments
                  memoryLimit : (nonnull NSNumber *) memoryLimit
                  assets : (NSString *) assets
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  NSNumber *id = @(++_id);

  BareKitModuleWorklet *worklet = [[BareKitModuleWorklet alloc] initWithModule:self
                                                                            id:id
                                                                      filename:filename
                                                                        source:source
                                                                     arguments:arguments
                                                                   memoryLimit:memoryLimit
                                                                        assets:assets];

  _worklets[id] = worklet;

  resolve(id);
}

RCT_EXPORT_METHOD(read : (nonnull NSNumber *) id
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  BareKitModuleWorklet *worklet = _worklets[id];

  if (worklet == nil) return reject(@"INVALID_ID", @"No such worklet found", nil);

  NSData *data = [worklet->_ipc read];

  if (data) return resolve([data base64EncodedStringWithOptions:0]);

  worklet->_ipc.readable = ^(BareIPC *ipc) {
    NSData *data = [ipc read];

    if (data == nil) return;

    ipc.readable = nil;

    resolve([data base64EncodedStringWithOptions:0]);
  };
}

RCT_EXPORT_METHOD(write : (nonnull NSNumber *) id
                  data : (NSString *) data
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  BareKitModuleWorklet *worklet = _worklets[id];

  if (worklet == nil) return reject(@"INVALID_ID", @"No such worklet found", nil);

  NSData *decoded = [[NSData alloc] initWithBase64EncodedString:data options:0];

  if ([worklet->_ipc write:decoded]) return resolve(nil);

  worklet->_ipc.writable = ^(BareIPC *ipc) {
    if ([ipc write:decoded]) {
      ipc.writable = nil;

      resolve(nil);
    }
  };
}

RCT_EXPORT_METHOD(suspend : (nonnull NSNumber *) id
                  linger : (nonnull NSNumber *) linger
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  BareKitModuleWorklet *worklet = _worklets[id];

  if (worklet == nil) return reject(@"INVALID_ID", @"No such worklet found", nil);

  [worklet _suspend:linger];

  resolve(nil);
}

RCT_EXPORT_METHOD(resume : (nonnull NSNumber *) id
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  BareKitModuleWorklet *worklet = _worklets[id];

  if (worklet == nil) return reject(@"INVALID_ID", @"No such worklet found", nil);

  [worklet _resume];

  resolve(nil);
}

RCT_EXPORT_METHOD(terminate : (nonnull NSNumber *) id
                  resolve : (RCTPromiseResolveBlock) resolve
                  reject : (RCTPromiseRejectBlock) reject) {
  BareKitModuleWorklet *worklet = _worklets[id];

  if (worklet == nil) return reject(@"INVALID_ID", @"No such worklet found", nil);

  [worklet _terminate];

  [_worklets removeObjectForKey:id];

  resolve(nil);
}

@end

File: android/src/main/java/to/holepunch/bare/kit/react/BareKitModule.java
package to.holepunch.bare.kit.react;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;
import to.holepunch.bare.kit.IPC;
import to.holepunch.bare.kit.Worklet;
import to.holepunch.bare.kit.react.NativeBareKitSpec;

public class BareKitModule extends NativeBareKitSpec {
  public static String NAME = "BareKit";

  private int id;
  private HashMap<Integer, BareKitModuleWorklet> worklets;

  BareKitModule(ReactApplicationContext context) {
    super(context);

    this.id = 0;
    this.worklets = new HashMap<>();
  }

  @Override
  public String
  getName() {
    return NAME;
  }

  @Override
  public void
  invalidate() {
    super.invalidate();

    for (BareKitModuleWorklet worklet : worklets.values()) {
      worklet.terminate();
    }

    worklets.clear();
  }

  @Override
  public double
  start(String filename, String source, ReadableArray arguments, double memoryLimit, String assets) {
    int id = ++this.id;

    ByteBuffer data;

    if (source == null) {
      data = null;
    } else {
      data = decode(source);
    }

    BareKitModuleWorklet worklet = new BareKitModuleWorklet(this, id, filename, data, arguments.toArrayList().toArray(new String[arguments.size()]), (int) memoryLimit, assets);

    this.worklets.put(id, worklet);

    return id;
  }

  @Override
  public void
  read(double id, Promise promise) {
    BareKitModuleWorklet worklet = worklets.get((int) id);

    if (worklet == null) {
      promise.reject("INVALID_ID", new Error("No such worklet found"));
      return;
    }

    read(worklet.ipc, promise);
  }

  private void
  read(IPC ipc, Promise promise) {
    ByteBuffer data = ipc.read();

    if (data != null) {
      ipc.readable(null);

      promise.resolve(encode(data));
    } else {
      ipc.readable(() -> read(ipc, promise));
    }
  }

  @Override
  public void
  write(double id, String data, Promise promise) {
    BareKitModuleWorklet worklet = worklets.get((int) id);

    if (worklet == null) {
      promise.reject("INVALID_ID", new Error("No such worklet found"));
      return;
    }

    write(worklet.ipc, decode(data), promise);
  }

  private void
  write(IPC ipc, ByteBuffer data, Promise promise) {
    if (ipc.write(data)) {
      ipc.writable(null);

      promise.resolve(null);
    } else {
      ipc.writable(() -> write(ipc, data, promise));
    }
  }

  @Override
  public void
  suspend(double id, double linger) {
    BareKitModuleWorklet worklet = worklets.get((int) id);

    if (worklet == null) return;

    worklet.suspend((int) linger);
  }

  @Override
  public void
  resume(double id) {
    BareKitModuleWorklet worklet = worklets.get((int) id);

    if (worklet == null) return;

    worklet.resume();
  }

  @Override
  public void
  terminate(double id) {
    BareKitModuleWorklet worklet = worklets.get((int) id);

    if (worklet == null) return;

    worklet.terminate();

    worklets.remove(worklet.id);
  }

  private String
  encode(ByteBuffer data) {
    return StandardCharsets.UTF_8.decode(Base64.getEncoder().encode(data)).toString();
  }

  private ByteBuffer
  decode(String data) {
    return ByteBuffer.wrap(Base64.getDecoder().decode(data));
  }

  private static class BareKitModuleWorklet {
    int id;
    BareKitModule module;
    Worklet worklet;
    IPC ipc;

    BareKitModuleWorklet(BareKitModule module, int id, String filename, ByteBuffer source, String[] arguments, int memoryLimit, String assets) {
      this.id = id;
      this.module = module;

      Worklet.Options options = new Worklet.Options().memoryLimit(memoryLimit).assets(assets);

      this.worklet = new Worklet(options);
      this.worklet.start(filename, source, arguments);

      this.ipc = new IPC(this.worklet);
    }

    void
    suspend(int linger) {
      worklet.suspend(linger);
    }

    void
    resume() {
      worklet.resume();
    }

    void
    terminate() {
      ipc.close();
      worklet.terminate();
    }
  }
}
